### ISSUE TITLE
Fine tuning on an unknown language (very bad results on LoRA) 

### ISSUE THREAD

**Issue by Anindyadeep** [Ucas-HaoranWei/GOT-OCR2.0#184]
Hi I was trying to fine-tune the model on an unknown language. The Language I was trying to do was [pashto](https://en.wikipedia.org/wiki/Pashto). Now I was aware that GOT-OCR was trained for this language. So I did this for the overall training pipeline. 

### Making a dataset

I saw that there was a [wikipedia dataset under pashto](https://huggingface.co/datasets/wikimedia/wikipedia/viewer/20231101.ps) language. So I downloaded the dataset and then used [Microsoft genalog](https://github.com/microsoft/genalog) to convert written text chunks of this pashto wikipedia datasets to images. 

Here is an example:
![image](https://github.com/user-attachments/assets/15be5049-8484-42bb-9be4-8826b13101a6)

And this is how the dataset looks like:

```json
[
  {
    "query": "\nOCR",
    "response": "یو له مدرنه تاریخ پوهانو څخه, چاچې د خپلو لویو ممکنه پروژو په مرسته د تاریخي لاسوندونه شتوالی ولري\n منځنۍ پېړۍ\n نوی پېر\n اوسمهال\n\nسرچينې \n\n History & Mathematics: Historical Dynamics and Development of Complex Societies. Moscow: KomKniga, 2006. ISBN 5-484-01002-0\n\nپېښليکچاپېريال پوهنه چې په انگريزي ژبلويې کچې د هغوی په چاپېريال کې بېلابېل بدلونونه څېړل ...",
    "images": [
      "text_chunks/chunk_0001.png"
    ]
  },
```

Now, I used this command to fine-tune the model using Lora using ms-swift

```bash
CUDA_VISIBLE_DEVICES=0,1,2,3 \
NPROC_PER_NODE=4 \
swift sft \
    --model_type got-ocr2 \
    --model_id_or_path stepfun-ai/GOT-OCR2_0 \
    --dataset "./training_data_got_cleaned.json" \
    --num_train_epochs 5 \
    --sft_type lora \
    --output_dir output \
    --max_length 4096 \
    --batch_size 4 \
    --learning_rate 1e-4 
```
It was giving around 70% accuracy using token matching as the eval. As a result I was quite hopeful. But then it turned out that, it gave me very bad result when I did testing for some test images. I am not sure why is the case. 

For example this was the type of output it was generating:

```txt
ar-ga'm, s.m. (1st) A. B. A belt, erection, perplexity, dilemma. A screen, shelter, concealment, protection. A fort, manacle, prison, place. Pl. m. 4. A fetter, manacle, restraint. Desire, wish, inclination. Also, 3. Adj. Obstructed, hindered, stopped, clogged, embarrassed, perplexed, emasculated. (Fem.) Alas I am sorrowful, miserable, wretched, distressed, afflicted, troubled, perplexed, emasculated. (Fem.) Alas I am sorrowful, miserable, wretched, distressed, afflicted, troubled, perplexed, emasculated. To trans. lie in wait to, to lie in ambush or to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be sorry for, to de- trans. To regret, to be
```

So am I missing something here? I am not sure why is the case. One reason could be that the test images might not be under the same distribution of the synthetic training data. But any pointers more than that are appreciated.

**Comment by Ucas-HaoranWei** [comment]
Hi, I think two points can improve the performance of other languages: 
1. Turn off Lora and fine-tune the decoder‘s full parameters; 
3. Unfreeze the encoder --- change the line132 and 146 'False' to True"
 https://github.com/Ucas-HaoranWei/GOT-OCR2.0/blob/main/GOT-OCR-2.0-master/GOT/model/GOT_ocr_2_0.py#L132.

**Comment by Anindyadeep** [comment]
> 2\. Unfreeze the encoder --- change the line132 and 146 'False' to True"

Hi, thanks for the reply, so basically 

```
# fine-tuning LLM & projector, freeze vision encoder
CUDA_VISIBLE_DEVICES=0 swift sft \
    --model_type got-ocr2 --model_id_or_path stepfun-ai/GOT-OCR2_0 \
  --sft_type lora \
  --dataset latex-ocr-print#5000
```

here in the `stepfun-ai/GOT-OCR2_0`, if I give a custom path where if I change [the model's](https://huggingface.co/stepfun-ai/GOT-OCR2_0/blob/main/modeling_GOT.py) lines from setting grad to True, then I am guessing it should work right?

**Comment by Anindyadeep** [comment]
> Hi, I think two points can improve the performance of other languages:
> 
>     1. Turn off Lora and fine-tune the decoder‘s full parameters;
> 
>     2. Unfreeze the encoder --- change the line132 and 146 'False' to True"
>        https://github.com/Ucas-HaoranWei/GOT-OCR2.0/blob/main/GOT-OCR-2.0-master/GOT/model/GOT_ocr_2_0.py#L132.

it worked, but it seems like the model did not learn anything, which is something very confusing, not sure is it the dataset or something else.

**Comment by thhung** [comment]
I also trained on the same fashion that you did above with the advices from the author but for a different language. My learning is stuck around 6x% but cannot increase. 

@Ucas-HaoranWei Do you have any recommendation to debug the problem in this case? My dataset is around 65k data points btw if this info is useful.

**Comment by Anindyadeep** [comment]
> I also trained on the same fashion that you did above with the advices from the author but for a different language. My learning is stuck around 6x% and cannot increase.
> 
> @Ucas-HaoranWei Do you have any recommendation to debug the problem in this case? My dataset is around 65k data points btw if this info is useful.

in my case, it is strictly getting overfitted in the training set and giving very bad results for any kind of picture taken from internet, which includes the writings of the language I trained the OCR model for.

**Comment by Ucas-HaoranWei** [comment]
I don't think it's the decoder's problem, it's the encoder's inability to distinguish the response of new language features, which causes the decoder to be unable to output the results correctly. Therefore, the new language must be trained from stage 2 or even 1 and requires a large amount of data

**Comment by Anindyadeep** [comment]
> I don't think it's the decoder's problem, it's the encoder's inability to distinguish the response of new language features, which causes the decoder to be unable to output the results correctly. Therefore, the new language must be trained from stage 2 or even 1 and requires a large amount of data

I mean I had around 100K sample, but may be the way I am generating the text -> image pairs might not be under the same distribution as the "actual" documents. I have using genalog for doing this dataset generation.

https://github.com/microsoft/genalog

**Comment by Ucas-HaoranWei** [comment]
Sorry, I reply to  @thtung~
What is the effect of the new image generated by the genalog？

**Comment by thhung** [comment]
@Ucas-HaoranWei My guess is the data doesn't look "real" like the images on your dataset. Do you have any plan to support other languages than English or Chinese? I guess you have computes to do that if you have data and manpower.

**Comment by Anindyadeep** [comment]
> @Ucas-HaoranWei My guess is the data doesn't look "real" like the images on your dataset. Do you have any plan to support other languages than English or Chinese? I guess you have computes to do that if you have data and manpower. This might help your work get more popular and be adopted by more people I guess.

Yes you are right on this
